mep3_navigation/src/distance_angle/distance_angle_regulator.cpp

```diff
@@ -0,0 +1,229 @@
+        "/odom", 100, std::bind(&DistanceAngleRegulator::odometry_callback, this, _1));
+    twist_publisher_ = this->create_publisher<geometry_msgs::msg::Twist>("/cmd_vel", 10);
+
+    command_subscription_ = this->create_subscription<mep3_msgs::msg::MotionCommand>("/da_command", 100, std::bind(&DistanceAngleRegulator::command_callback, this, _1));
+
+    this->declare_parameter("kp_distance", 25000.0);
+    this->declare_parameter("ki_distance", 1.5);
+    this->declare_parameter("kd_distance", 1120.0);
+
+    this->declare_parameter("kp_angle", 40000.0);
+    this->declare_parameter("ki_angle", 22.0);
+    this->declare_parameter("kd_angle", 1500.0);
+
+    debug_ = this->declare_parameter("debug", true);
+    parameters_callback_handle_ = this->add_on_set_parameters_callback(std::bind(&DistanceAngleRegulator::parameters_callback, this, std::placeholders::_1));
+
+    this->get_parameter("kp_distance", regulator_distance_.kp);
+    this->get_parameter("ki_distance", regulator_distance_.ki);
+    this->get_parameter("kd_distance", regulator_distance_.kd);
+    regulator_distance_.clamp_min = -2000.0;
+    regulator_distance_.clamp_max = 2000.0;
+    regulator_distance_.integrator_min = -800.0;
+    regulator_distance_.integrator_max = 800.0;
+
+    this->get_parameter("kp_angle", regulator_angle_.kp);
+    this->get_parameter("ki_angle", regulator_angle_.ki);
+    this->get_parameter("kd_angle", regulator_angle_.kd);
+    regulator_angle_.clamp_min = -5000.0;
+    regulator_angle_.clamp_max = 5000.0;
+    regulator_angle_.integrator_min = -1500.0;
+    regulator_angle_.integrator_max = 1500.0;
+
+    robot_distance_ = 0;
+    position_initialized_ = false;
+
+    distance_profile_ = MotionProfile(0, 0.1, 0.02);
+    angle_profile_ = MotionProfile(0, 0.5, 0.25);
+}
+
+void DistanceAngleRegulator::odometry_callback(const nav_msgs::msg::Odometry::SharedPtr msg)
+{
+    rclcpp::Time time = this->get_clock()->now();
+
+    double robot_x = msg->pose.pose.position.x;
+    double robot_y = msg->pose.pose.position.y;
+
+    if (!position_initialized_)
+    {
+        prev_robot_x_ = robot_x;
+        prev_robot_y_ = robot_y;
+        position_initialized_ = true;
+
+        tf2::Quaternion q(
+            msg->pose.pose.orientation.x,
+            msg->pose.pose.orientation.y,
+            msg->pose.pose.orientation.z,
+            msg->pose.pose.orientation.w);
+        tf2::Matrix3x3 m(q);
+        double roll, pitch, yaw;
+        m.getRPY(roll, pitch, yaw);
+        robot_angle_ = yaw;
+
+        distance_profile_.plan(0, 0, 0, 0, time);
+        angle_profile_.plan(robot_angle_, robot_angle_, 0, 0, time);
+    }
+
+    tf2::Quaternion q(
+        msg->pose.pose.orientation.x,
+        msg->pose.pose.orientation.y,
+        msg->pose.pose.orientation.z,
+        msg->pose.pose.orientation.w);
+    tf2::Matrix3x3 m(q);
+    double roll, pitch, yaw;
+    m.getRPY(roll, pitch, yaw);
+    robot_angle_ = yaw;
```

Comment:
S obzirom da imamo rotaciju samo oko z ose (yaw) Euler ugao bi trebalo da je jednak `2 asin (q_z)` (molim te provjeri):
```suggestion
    robot_angle_ = 2 * std::asin(q.z);
```

Reply:
Ne moze, menjaju se `z` i `w`

Reply:
Tvoje novo rjesenje je mnogo bolje. @angstrem98 mozes im ostaviti odgovor na ros.answers :)